src/ProjectOrigin.Electricity.Server/Services/ProtoDeserializer.cs

```diff
@@ -27,7 +29,20 @@ public ProtoDeserializer(Assembly assembly)
 
     public IMessage Deserialize(string type, ByteString content)
     {
-        var descriptor = _typeDictionary[type];
-        return descriptor.Parser.ParseFrom(content);
+        if (_typeDictionary.TryGetValue(type, out var descriptor))
+        {
+            try
+            {
+                return descriptor.Parser.ParseFrom(content);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidPayloadException($"Could not deserialize invalid payload of type ”{type}”", ex);
+            }
+        }
+        else
+        {
+            throw new InvalidPayloadException($"Could not deserialize unknown type ”{type}”");
+        }
```

Comment:
Minor thing, but this could be a one-liner:

```suggestion

        throw new InvalidPayloadException($"Could not deserialize unknown type ”{type}”");
```